### DISCUSSION TITLE
GDS Errors counter increases

### DISCUSSION THREAD

**Discussion by Gonta01** [nasa/fprime#2876] [Q&A]
From time to time GDS error increases.
Does somebody know what is possible reason (maybe UART implementation specific to hardware?) and how to debug GDS errors:
![image](https://github.com/user-attachments/assets/f358e4d4-6c43-48d9-a825-ca9f68aa673c)

**Comment by Gonta01** [comment]
Also in console the following message can be seen:
![image](https://github.com/user-attachments/assets/e3403bec-f4de-4e2f-a365-59ffbcf48f40)

**Reply by LeStarch** [reply]
GDS errors occur in the GDS subsystem.  The most common source of GDS errors is when your backend (Python) stops talking to your front end (Web Browser).  These errors are logged at the bottom of the "Advanced" tab.  Have a look there for more details.

**Reply by LeStarch** [reply]
The checksum validation is a different error.  It means the GDS received the bytes that look like a message, but did not pass checksum.  i.e.

0xdeadbeef (size token) ... bytes ... (checksum token) ...but the recalculation of the checksum failed.  This could be caused if:

1. You send 0xdeadbeef in your data payload, and got desynced from a frame
2. You dropped bytes between flight and ground
3. You switched framing protocols without updating the GDS.

**Reply by Gonta01** [reply]
In Advanced tab is see such type of error:
![image](https://github.com/user-attachments/assets/6af62cc4-259a-4b79-9c77-b7e005ee0f97)

**Reply by LeStarch** [reply]
This warns that the polling code that pulls channels from the backend to the frontend is falling behind its expected interval.  This is usually caused by one of two things:

1. Network latency between the UI and the backend
2. Large data transfers that take longer then expected

If these errors are occasional, it is ok to clear them.  If these errors are constant, then some tuning is likely.

Tuning:
- Make sure there is good network connection between the host running `fprime-gds` and the machine running the browser.
- Try decreasing the number of channels requested at a time (Advanced Tab)
- Try decreasing the polling interval of channels (Advanced Tab)
- As a last resort, reduce the number of channels being downlinked

Tuning explanation:

At a polling interval configured on the Advanced Tab, the frontend UI asks the backend data store for new channels.  The total number that can be transferred is limited by another Advanced Tab configuration setting.  The above error happens when a previous request is still running when it is time to ask for the next set of data.

When intermittent, this error reveals occasional slowness in the network between the `fprime-gds` and the associated client browser.  When continuous, this error reveals there is too much data to transfer over in the given interval.  Smaller more frequent requests typically out-perform larger less frequent ones (or so it seems).  So decreasing the max channel count and the interval may help.
